datafusion/core/src/datasource/file_format/parquet.rs

```diff
@@ -1015,15 +1050,21 @@ async fn concatenate_parallel_row_groups(
     while let Some(task) = serialize_rx.recv().await {
         let result = task.join_unwind().await;
         let mut rg_out = parquet_writer.next_row_group()?;
-        let (serialized_columns, _cnt) = result?;
+        let (serialized_columns, mut rg_reservation, _cnt) = result?;
         for chunk in serialized_columns {
             chunk.append_to_row_group(&mut rg_out)?;
+            rg_reservation.free();
+
             let mut buff_to_flush = merged_buff.buffer.try_lock().unwrap();
+            file_reservation.try_resize(buff_to_flush.len())?;
+
             if buff_to_flush.len() > BUFFER_FLUSH_BYTES {
                 object_store_writer
                     .write_all(buff_to_flush.as_slice())
                     .await?;
+                rg_reservation.shrink(buff_to_flush.len());
```

Comment:
Since `file_reservation` seems to now be tracking `buff_to_flush`, not `rg_reservation` I don't think `rg_reservation` needs to be changed on flush.

Also, `rg_reservation.free()` is called a few lines above, it should already be zero

Reply:
TY. I caught this, and two other cleanup tasks. Pushed.